### PR TITLE
Abstract out creation "Of" and conversion "To" Quantity members

### PR DESCRIPTION
**Pull request by atmoos**
- All quantities had "Copy Paste" versions of To{...} and Create{...} methods.
- Now replaced by reusable structs
- Held together by IFactory<TQuantity, TDimension> interface
- Performance has degraded, but
- Readability has increased (this has higher importance)

Note: Naming is still bad in many (new) areas. Ought to change when I publish this PR.

Initially borne out of an idea to make serialization easier...

**In a nutshell**

Previously:
```csharp
Length smallMetric = Length.Si<Micro, Metre>(23);
```

Now:
```csharp
Length smallMetric = Length.Of(23).Si<Micro, Metre>();
```

### CODE REVIEW

File: quantities/factories/IFactory.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+namespace Quantities.Factories;
+
+public interface IFactory { /* marker interface */ }
+
+public interface IFactory<in TFactory, out TTo, out TCreate> : IFactory
```
Comment (atmoos) [inline quantities/factories/IFactory.cs:5]: This is a "pseudo" interface that "defines" how to "spell" the create (`Of`) and convert (`To`) members.

What `To` and `Of` actually return is implementation specific.

Comment (atmoos) [reply quantities/factories/IFactory.cs:5]: Note also that most enabling types are in the folder "factories".

File: quantities.benchmark/CreateQuantities.cs [hunk 2]
```diff
@@ -1,64 +1,71 @@
...
+  DefaultJob : .NET 7.0.3 (7.0.323.12801), X64 RyuJIT AVX2
 
-|         Method |       Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
-|--------------- |-----------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
-|   CreateObject |  6.6147 ns | 0.0908 ns | 0.0805 ns | 1.000 |    0.00 | 0.0057 |     - |     - |      24 B |
-| CreateObjectIn |  6.0880 ns | 0.0495 ns | 0.0463 ns | 0.921 |    0.01 | 0.0057 |     - |     - |      24 B |
-|   CreateStruct |  0.0547 ns | 0.0147 ns | 0.0138 ns | 0.008 |    0.00 |      - |     - |     - |         - |
-|   CreateLength |  4.2311 ns | 0.0568 ns | 0.0504 ns | 0.640 |    0.01 |      - |     - |     - |         - |
-| CreateVelocity | 10.0543 ns | 0.0173 ns | 0.0135 ns | 1.522 |    0.02 | 0.0057 |     - |     - |      24 B |
+|                       Method |      Mean |     Error |    StdDev | Ratio | RatioSD |   Gen0 | Allocated | Alloc Ratio |
```
Comment (atmoos) [inline quantities.benchmark/CreateQuantities.cs:63]: Notice that creation has remained constant for simple types such as ´Length´. But creating more complex quantities, such as `Velocity` is now twice as slow and still allocates :-(

**Review by RaduTerec — APPROVED** [review]
Most questions are questions about understanding what you did. Stuff like:
- when do you use Create and when an internal constructor
- when do you use struct and when class
- why do some quants have a root while others don't.

I hope that the next reviews will be smaller because I wish to dive deep in understanding what you actually built here 😁

File: README.md [hunk 3]
```diff
@@ -51,33 +51,33 @@ Usage is designed to be intuitive using:
...
 ```
 
 ### Conversion
 
 ```csharp
-Length miles = metres.ToImperial<Mile>();
-Length kilometres = metres.ToSi<Kilo, Metre>();
-Velocity metresPerSecond = kilometresPerHour.To<Metre>().PerSecond();
-Velocity milesPerHour = kilometresPerHour.ToImperial<Mile>().Per<Hour>();
+Length miles = metres.To.Imperial<Mile>();
```
[the remark this answers is not in the document]
Comment (atmoos) [reply README.md:63]: Thanks :-)

File: quantities.benchmark/CreateQuantities.cs [hunk 4]
```diff
@@ -1,64 +1,71 @@
...
+    [MethodImpl(MethodImplOptions.NoOptimization)]
+    public DummyStruct CreateStruct() => new(in this.value);
 
     [Benchmark]
-    public DummyObject CreateObjectIn() => new DummyObject(in this.value);
+    public Length CreateFastQuantity() => Length.Of(in this.value).Si<Kilo, Metre>();
 
     [Benchmark]
-    public DummyStruct CreateStruct() => new DummyStruct(this.value);
+    [MethodImpl(MethodImplOptions.NoOptimization)]
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities.benchmark/CreateQuantities.cs:44]: :-)

File: quantities.benchmark/CreateQuantities.cs (hunk 2 above)
Comment (atmoos) [reply quantities.benchmark/CreateQuantities.cs:63]: You are absolutely right. We're already fast. I did set it as a goal for these types to be blazingly fast as a challenge and as I see them as a drop in for `double`.
Let's take this into more balanced consideration as the project evolves.

File: quantities/factories/CubicFactory.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+using Quantities.Measures;
+using Quantities.Prefixes;
+using Quantities.Units;
+using Quantities.Units.Imperial;
+using Quantities.Units.NonStandard;
+using Quantities.Units.Si;
+
+namespace Quantities.Factories;
+
+public readonly struct CubicFactory<TQuantity, TCompound, TCubic, TLinear> : ICubicFactory<TQuantity, TCubic, TLinear>
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/CubicFactory.cs:11]: Sure, let's do that this weekend :-)

File: quantities/factories/Factories.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
+using Quantities.Dimensions;
+using Quantities.Prefixes;
+using Quantities.Units.Imperial;
+using Quantities.Units.NonStandard;
+using Quantities.Units.Si;
+
+namespace Quantities.Factories;
+
+public interface ISiFactory<out TQuantity, in TDimension> : IFactory
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/Factories.cs:9]: Yes, they could. But they couldn't make any proper use of it. Reason being is that all quantity types have internal constructors by design.

File: quantities/factories/Factories.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    public TQuantity Binary<TPrefix, TUnit>() where TPrefix : IBinaryPrefix where TUnit : IMetricUnit, TDimension;
+}
+
+public interface IImperialFactory<out TQuantity, in TDimension>
+    where TDimension : IDimension
+{
+    public TQuantity Imperial<TUnit>() where TUnit : IImperialUnit, TDimension;
+}
+
+public interface INonStandardFactory<out TQuantity, in TDimension>
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/Factories.cs:35]: Will fix

File: quantities/factories/PerFactory.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+using Quantities.Prefixes;
+using Quantities.Units.Imperial;
+using Quantities.Units.NonStandard;
+using Quantities.Units.Si;
+using IDimension = Quantities.Dimensions.IDimension;
+
+namespace Quantities.Factories;
+
+public readonly struct Denominator<TFactory>
+    where TFactory : ICreatable<TFactory>, IFactory
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/PerFactory.cs:12]: Actually, there is no denominator here. This is only a factory that enables the fluent api of say a velocity.

```csharp
Velocity.Of(4).Si<Kilo, Metre>().Per.Metric<Hour>();
```
where:

- The value of four is nether the value of the nominator, nor the denominator. Its the value taken as _the_ velocity.
- The nominator will hold the "length", and
- The denominator the "time" types

So, the Denomiator and Nominator do not hold values, they hold types. Hope this makes sense?

File: quantities/factories/PerFactory.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+{
+    private readonly ICreate creator;
+    public TFactory Per => TFactory.Create(in this.creator);
+    internal Denominator(in ICreate builder) => this.creator = builder;
+}
+
+public readonly struct Nominator<TCreate, TNominator, TFactory> : ICompoundFactory<Denominator<TFactory>, TNominator>
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/PerFactory.cs:19]: Please see the comment regarding the denominator and division by zero.

File: quantities/quantities/Area.cs [hunk 10]
```diff
@@ -1,98 +1,25 @@
...
-using Quantities.Prefixes;
-using Quantities.Units;
-using Quantities.Units.Imperial;
-using Quantities.Units.NonStandard;
-using Quantities.Units.Si;
 
 namespace Quantities.Quantities;
 
-public readonly struct Area : IQuantity<Area>, IArea<Length>
+public readonly struct Area : IQuantity<Area>, IArea<ILength>
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/quantities/Area.cs:9]: Thanks, I'm actually not really happy with it. As `Area` rally _is_ an `IArea<Length>`, not an `IArea<ILength>` .
But this is somewhat philosophical. So I'll take the flowers :-)

File: quantities/quantities/ElectricalResistance.cs [hunk 11]
```diff
@@ -1,51 +1,35 @@
...
 namespace Quantities.Quantities;
 
 public readonly struct ElectricalResistance : IQuantity<ElectricalResistance>, IElectricalResistance
-    , ISiDerived<ElectricalResistance, IElectricalResistance>
+    , IFactory<ElectricalResistance>
+    , IFactory<ISiFactory<ElectricalResistance, IElectricalResistance>, SiTo<ElectricalResistance, IElectricalResistance>, SiCreate<ElectricalResistance, IElectricalResistance>>
     , IMultiplyOperators<ElectricalResistance, ElectricCurrent, ElectricPotential>
 {
-    private static readonly Creator create = new();
+    private static readonly IRoot root = new SiRoot<Ohm>();
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/quantities/ElectricalResistance.cs:16]: Thanks :-)

File: quantities/quantities/Energy.cs [hunk 12]
```diff
@@ -1,82 +1,24 @@
...
 namespace Quantities.Quantities;
 
 public readonly struct Energy : IQuantity<Energy>, IEnergy<Mass, Length, Time>
-    , ISiDerived<Energy, IEnergy>
-    , IImperial<Energy, IEnergy>
+    , IFactory<ICompoundFactory<Energy, IEnergy<Mass, Length, Time>>, Energy.Factory<LinearTo>, Energy.Factory<LinearCreate>>
     , IDivisionOperators<Energy, Time, Power>
     , IDivisionOperators<Energy, Power, Time>
 {
     private readonly Quant quant;
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/quantities/Energy.cs:17]: You're quite right. The "root" concept is - as of now - more an Idea that I planted in some quantity types, rather than a full blown concept.
What I don't like is that it introduces a dependency to units. I'd rather the quantities didn't know about any concrete unit at all.

A design question we should try and solve together. See also issue #10.

File: quantities/units/Si/Derived/Newton.cs [hunk 13]
```diff
@@ -2,7 +2,7 @@
 
 namespace Quantities.Units.Si.Derived;
 
-public readonly struct Newton : ISiDerivedUnit, IForce
+public readonly struct Newton : ISiUnit, IForce
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/units/Si/Derived/Newton.cs:5]: IMHO the different namespaces are differentiation enough.
I try to use the names defined either by a governing body (SI) or by the consensus on wikipedia.
I should start to consistently add links to the respective Wikipedia article for every unit.
Also, I want to avoid having to "invent" new names, as there are way too many units for which we'd need to come up with naming policies. Namespaces are - as of now - the right way to resolve naming conflicts.

The other "Newton"'s namespace is IMHO distinct enough compared to this Newton's namespace:
`Quantities.Units.NonStandard.Temperature` vs `Quantities.Units.Si.Derived`, respectively.

File: quantities/factories/PerFactory.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (RaduTerec) [reply quantities/factories/PerFactory.cs:12]: Yes, it does make sense. It also makes sense to leave the handling of a negative infinity to the user. I would still document this, just like the negative value for distance.

File: quantities.benchmark/AddingQuantities.cs [hunk 14]
```diff
@@ -11,11 +11,11 @@ namespace Quantities.Benchmark;
 [MemoryDiagnoser]
 public class AddingQuantities
 {
-    private Length sameMetric = Length.Si<Kilo, Metre>(-7);
-    private Length largeMetric = Length.Si<Kilo, Metre>(3);
-    private Length smallMetric = Length.Si<Micro, Metre>(23);
-    private Length largeImperial = Length.Imperial<Mile>(-3);
-    private Length smallImperial = Length.Imperial<Inch>(55);
+    private Length sameMetric = Length.Of(-7).Si<Kilo, Metre>();
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities.benchmark/AddingQuantities.cs:14]: Done

File: quantities/factories/Factories.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    public TQuantity Metric<TPrefix, TUnit>() where TPrefix : IMetricPrefix where TUnit : IMetricUnit, TDimension;
+}
+
+public interface IBinaryFactory<out TQuantity, in TDimension> : IFactory
+    where TDimension : IDimension
+{
+    public TQuantity Binary<TPrefix, TUnit>() where TPrefix : IBinaryPrefix where TUnit : IMetricUnit, TDimension;
+}
+
+public interface IImperialFactory<out TQuantity, in TDimension>
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/Factories.cs]: Done

File: quantities/factories/Factories.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/Factories.cs]: Done

File: quantities/factories/PerFactory.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/PerFactory.cs:12]: Done.

File: quantities/units/Imperial/Mass/Ton.cs [hunk 16]
```diff
@@ -3,7 +3,7 @@
 
 namespace Quantities.Units.Imperial.Mass;
 
-public readonly struct Ton : IImperial, IMass
+public readonly struct Ton : IImperialUnit, IMass
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/units/Imperial/Mass/Ton.cs:6]: The world of units is quite fun :-)

File: quantities/units/Si/Derived/Newton.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/units/Si/Derived/Newton.cs:5]: Done.

File: quantities/factories/PowerFactory.cs [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    where TCreate : ICreate, IAliasingCreate
+    where TLinear : Dimensions.IDimension, ILinear
+    where TQuantity : IFactory<TQuantity>
+{
+    private readonly TCreate creator;
+    internal PowerFactory(in TCreate value) => this.creator = value;
+    public TQuantity Si<TUnit>() where TUnit : ISiUnit, TLinear => TQuantity.Create(this.creator.Create<Si<TUnit>>());
+    public TQuantity Si<TPrefix, TUnit>()
+        where TPrefix : IMetricPrefix
+        where TUnit : ISiUnit, TLinear => TQuantity.Create(this.creator.Create<Si<TPrefix, TUnit>>());
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/PowerFactory.cs:20]: I'm not quite sure what you mean here.
There is no new constraint in the `PerFactory`, there are only new statements that create instances.

Please re-open if I misunderstood.

File: quantities/measures/Build.cs
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/measures/Build.cs:1]: You're quite right. We discussed this a bit today.

As discussed I added some more documentation to "Design.md" and with issue #11 I'll be adding a more documentation down the road.

File: quantities/measures/transformations/Divide.cs [hunk 18]
```diff
@@ -1,13 +1,13 @@
 namespace Quantities.Measures.Transformations;
 
-internal sealed class Divide : ICreate<ICreate<Quant>>
+internal sealed class Divide : IInject<IInject<Quant>>
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/measures/transformations/Divide.cs:3]: Sure.
First, these are all helper classes that are used to transform quants from one dimension to another. As such all instances of these transformations are static fields in the Quantity classes (when needed), which means their allocations are irrelevant during runtime (their only instantiated once per Quantity). The fields are all declared using the interface `IInject<T>`, so if I'd be using structs boxing conversion would happen (also only once).

Therefore, it is technically not relevant whether we use structs or classes, so it's only a matter of taste, essentially.

PS: There is a minor argument that could act in favour of classes. Should we ever need to declare the fields using the actual type not the interface type. The calls remain allocation free if we use classes (counter-intuitive, I know....). Whereas if we'd be using structs, we'd be incurring a boxing conversion (and hence an allocation) every time we use a transformation.

PPS: This points to a quirk of C# where not the usage of the type determines allocations but the design of the type. Here, it has the arguably bad side effect, that the usage of a type defines the design of a type :-(

[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/measures/transformations/Divide.cs:3]: A topic for a philosophy session actually :-)

File: quantities/measures/Build.cs [hunk 19]
```diff
@@ -12,7 +12,7 @@ internal static class Build<TMeasure> where TMeasure : IMeasure
     public static Quant With<TInjector>(in Double value)
         where TInjector : IInjector, new()
     {
-        return new(value, in MapPool<TInjector>.Item);
+        return new(in value, in MapPool<TInjector>.Item);
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/measures/Build.cs:15]: You mean instead of `With`?

I think I started out like that when the class was still called "Builder". Usage then looked like this:

```csharp
Builder<TMeasure>.Create<TInjector>(42);
```
which IMHO looks super weird with those type parameters.

Compare this to:

```csharp
Build<TMeasure>.With<TInjector>(42);
``` 
Much closer to an English sentence. What do you think?

IMHO the only remaining weirdness is the value we pass at the very end....

File: quantities/factories/LinearFactories.cs [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+        where TUnit : ISiUnit, TDimension => TQuantity.Create(this.value.To<Si<TPrefix, TUnit>>());
+    public TQuantity Metric<TUnit>() where TUnit : IMetricUnit, TDimension => TQuantity.Create(this.value.To<Metric<TUnit>>());
+    public TQuantity Metric<TPrefix, TUnit>()
+        where TPrefix : IMetricPrefix
+        where TUnit : IMetricUnit, TDimension => TQuantity.Create(this.value.To<Metric<TPrefix, TUnit>>());
+    public TQuantity Imperial<TUnit>() where TUnit : IImperialUnit, TDimension => TQuantity.Create(this.value.To<Imperial<TUnit>>());
+    public TQuantity NonStandard<TUnit>() where TUnit : INoSystemUnit, TDimension => TQuantity.Create(this.value.To<NonStandard<TUnit>>());
+}
+
+public readonly struct LinearFactory<TQuantity, TDimension> : ICompoundFactory<TQuantity, TDimension>, ICreatable<LinearFactory<TQuantity, TDimension>>
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/LinearFactories.cs:46]: Discussed.

File: quantities.test/LengthTest.cs [hunk 21]
```diff
@@ -143,9 +150,9 @@ public void LengthByDivisionIsEqualToLengthByConstruction()
     [Fact]
     public void SiLengthBySiTimeIsVelocity()
     {
-        Length distance = Length.Si<Milli, Metre>(100);
-        Time duration = Time.Seconds(20);
-        Velocity expected = Velocity.Si<Milli, Metre>(5).PerSecond();
+        Length distance = Length.Of(100).Si<Milli, Metre>();
+        Time duration = Time.Of(20).Si<Second>();
+        Velocity expected = Velocity.Of(5).Si<Milli, Metre>().Per.Si<Second>();
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities.test/LengthTest.cs:155]: Added this to issue #11.

File: quantities/Extensions.cs [hunk 22]
```diff
@@ -1,58 +1,26 @@
...
-       where TUnit : ISiBaseUnit, ITime => new(builder.By<Si<TPrefix, TUnit>>());
-    public static DataRate PerSecond(this IBuilder<DataRate> builder) => new(builder.By<Si<Second>>());
-    internal static Quant To<TDim, TMeasure>(this Double value)
-      where TDim : Measures.IDimension where TMeasure : IMeasure, ILinear
+    internal static Quant To<TDim, TMeasure>(this in Double value)
+      where TDim : Measures.IDimension where TMeasure : IMeasure
       => Build<Power<TDim, TMeasure>>.With<Linear<TMeasure>>(in value);
-    internal static Quant To<TDim, TMeasure>(this Quant value)
-      where TDim : Measures.IDimension where TMeasure : IMeasure, ILinear
+    internal static Quant As<TDim, TMeasure>(this in Quant value)
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/Extensions.cs:11]: I think we discussed this last week. Resolving.

File: quantities/IFactory.cs [hunk 23]
```diff
@@ -0,0 +1,8 @@
+using Quantities.Measures;
+
+namespace Quantities;
+
+public interface IFactory<out TResult>
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/IFactory.cs:5]: Yes, purely a result of introducing "Of" and "To".

File: quantities/common/Scaling.cs [hunk 24]
```diff
@@ -1,4 +1,5 @@
 ﻿namespace Quantities;
 
-public interface IScaleUp { /* marker interface */ }
-public interface IScaleDown { /* marker interface */ }
+public interface IScale { /* marker interface */ }
```
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/common/Scaling.cs:3]: I think we've also discussed this last week. Resolving.

File: quantities/factories/CubicFactory.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (atmoos) [reply quantities/factories/CubicFactory.cs:11]: Thanks for the pointer to the custom attributes though!

File: quantities/measures/Build.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (RaduTerec) [reply quantities/measures/Build.cs:15]: I was wondering why you didn't use `Create` instead of `new`. I didn't understand at the time when you use the factory to create a new instance and when you use a simple instantiation. Now that I think of it, it's still a bit fuzzy 😅
